ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -533,10 +533,31 @@ PrjFS_Result PrjFS_DeleteFile(
     }
 
     // TODO(Mac): Ensure that races with hydration are handled properly
-    // TODO(Mac): Ensure file is not full before proceeding
     
     char fullPath[PrjFSMaxPath];
     CombinePaths(s_virtualizationRootFullPath.c_str(), relativePath, fullPath);
+    
+    struct stat path_stat;
+    stat(fullPath, &path_stat);
+    if (!(S_ISREG(path_stat.st_mode) || S_ISDIR(path_stat.st_mode)))
+    {
+        // Only files and directories can be deleted with PrjFS_DeleteFile
+        // Anything else should be treated as a full
+        *failureCause = PrjFS_UpdateFailureCause_FullFile;
+        return PrjFS_Result_EVirtualizationInvalidOperation;
+    }
+    
+    if (S_ISREG(path_stat.st_mode))
+    {
+        // TODO(Mac): Determine if we need a similar check for directories as well
+        PrjFSFileXAttrData xattrData = {};
+        if (!GetXAttr(fullPath, PrjFSFileXAttrName, sizeof(PrjFSFileXAttrData), &xattrData))
+        {
+            *failureCause = PrjFS_UpdateFailureCause_FullFile;
+            return PrjFS_Result_EVirtualizationInvalidOperation;
```

Comment:
I don't think this is an invalid operation. This can easily happen because of a race that causes a placeholder file to get converted to full right as the provider is trying to delete it.

Reply:
I personally think that's a mistake. An "invalid operation" should indicate that the caller of the API did something wrong, i.e. they have a bug. This is a situation that is totally unavoidable from the provider's perspective, so it's a runtime error, not a bug.